Mathlib/RingTheory/Flat/FaithfullyFlat.lean

```diff
@@ -343,6 +351,137 @@ end fixed_universe
 
 end exact
 
+section linearMap
+
+/--
+If `M` is a faithfully flat module, then for all linear maps `f`, the map `id ⊗ f = 0`, if and only
+if  `f = 0`. -/
+lemma zero_iff_lTensor_zero [h: FaithfullyFlat R M]
+    {N : Type*} [AddCommGroup N] [Module R N]
+    {N' : Type*} [AddCommGroup N'] [Module R N'] (f : N →ₗ[R] N') :
+    f = 0 ↔  LinearMap.lTensor M f = 0 :=
+  ⟨fun hf => hf.symm ▸ LinearMap.lTensor_zero M, fun hf => by
+    have := lTensor_reflects_exact R M f LinearMap.id (by
+      rw [LinearMap.exact_iff, hf, LinearMap.range_zero, LinearMap.ker_eq_bot]
+      apply Module.Flat.lTensor_preserves_injective_linearMap
+      exact fun _ _ h => h)
+    ext x; simpa using this (f x)⟩
+
+
+/--
+If `M` is a faithfully flat module, then for all linear maps `f`, the map `f ⊗ id = 0`, if and only
+if  `f = 0`. -/
+lemma zero_iff_rTensor_zero [h: FaithfullyFlat R M]
+    {N : Type*} [AddCommGroup N] [Module R N]
+    {N' : Type*} [AddCommGroup N'] [Module R N']
+    (f : N →ₗ[R] N') :
+    f = 0 ↔  LinearMap.rTensor M f = 0 :=
+  zero_iff_lTensor_zero R M f |>.trans
+  ⟨fun h => by ext n m; exact (TensorProduct.comm R N' M).injective <|
+    (by simpa using congr($h (m ⊗ₜ n))), fun h => by
+    ext m n; exact (TensorProduct.comm R M N').injective <| (by simpa using congr($h (n ⊗ₜ m)))⟩
+
+
+/--
+An `R`-module `M` is faithfully flat iff it is flat and for all linear maps `f`, the map
+`id ⊗ f = 0`, if and only if `f = 0`. -/
+lemma iff_zero_iff_lTensor_zero :
+    FaithfullyFlat R M ↔
+    (Module.Flat R M ∧
+      (∀ {N : Type max u v} [AddCommGroup N] [Module R N]
+        {N' : Type max u v} [AddCommGroup N'] [Module R N']
+        (f : N →ₗ[R] N'), f.lTensor M = 0 ↔ f = 0)):=
+  ⟨fun fl => ⟨inferInstance, fun f => zero_iff_lTensor_zero R M f |>.symm⟩,
+    fun ⟨flat, Z⟩ => iff_flat_and_lTensor_reflects_triviality R M |>.2 ⟨flat, fun N _ _ _ => by
+      have := Z (LinearMap.id : N →ₗ[R] N) |>.1 (by ext; exact Subsingleton.elim _ _)
+      rw [subsingleton_iff_forall_eq 0]
+      exact fun y => congr($this y)⟩⟩
+
+/--
+An `R`-module `M` is faithfully flat iff it is flat and for all linear maps `f`, the map
+`id ⊗ f = 0`, if and only if `f = 0`. -/
+lemma iff_zero_iff_rTensor_zero :
```

Comment:
Do we have the `→` implication using `Type*`?

Reply:
I suspect not, because the `→` direction uses `Module.FaithfullyFlat.iff_flat_and_rTensor_faithful` which we use the quotient `R / maximal ideal`, so the universe needs to be at least as large as `u` (the universe of `R`), and somewhere else we need `Module.Flat.iff_rTensor_exact` which need the universe level to be at least as large as `v` (the universe of the module)

Reply:
Basically if something is implied by faithful flatness then we can have universe polymorphism. But if we want to conclude some module as faithfully flat, then the testing object needs to be constrained by universe.

Reply:
Yes, this is also what I mean by the `→` implication (if the module if faithfully flat then blah blah).

Reply:
In that case, `Module.FaithfullyFlat. zero_iff_rTensor_zero` already exists some lines above this lemma

Reply:
OK, sorry.